typing/types.ml

```diff
@@ -1042,6 +1042,11 @@ let rec best_effort_compare_type_expr te1 te2 =
       | Tvariant _
       | Tpackage (_, _)
       | Tarrow (_, _, _, _)
+          (* NOTE: we can actually see Tsubst here in certain cases, eg during
+             [Ctype.copy] when copying the types inside of with_bounds. We also can't
+             compare Tsubst structurally, because the Tsubsts that are created in
+             Ctype.copy are cyclic (?). So the best we can do here is compare by id. *)
+      | Tsubst (_, _)
```

Comment:
As discussed today, I'm pretty sure this is wrong. (Just leaving this comment here so the wrongness doesn't get lost.)

Reply:
added more commentary about the wrongness